File: prototype/rails/src/test/domain/payments/payment_status_test.rb
require "test_helper"

module Domain
  module Payments
    class PaymentStatusTest < ActiveSupport::TestCase
      test "all names every status" do
        assert_equal %w[approved declined], PaymentStatus::ALL
      end

      test "an approved card records an approved payment" do
        assert_equal PaymentStatus::APPROVED, PaymentStatus.from_card_decision(CardDecision.approved("4242"))
      end

      test "a declined card records a declined payment" do
        decision = CardDecision.declined("0002", DeclineReason::GENERIC_DECLINE)

        assert_equal PaymentStatus::DECLINED, PaymentStatus.from_card_decision(decision)
      end
    end
  end
end

File: app/src/app/Domain/Payments/PaymentStatusTest.php
<?php

declare(strict_types=1);

namespace App\Domain\Payments;

it('records payment status from a card decision', function (CardDecision $decision, PaymentStatus $expected): void {
    expect(PaymentStatus::fromCardDecision($decision))->toBe($expected);
})->with([
    'an approved card records an approved payment' => [CardDecision::approved('4242'), PaymentStatus::Approved],
    'a declined card records a declined payment' => [CardDecision::declined('0002', DeclineReason::GenericDecline), PaymentStatus::Declined],
]);
